asmcomp/amd64/arch.ml

```diff
@@ -13,13 +13,25 @@
 (*                                                                        *)
 (**************************************************************************)
 
+(* LZCNT instruction is not avialble on Intel Architecures prior to Haswell.
+
+   Important: lzcnt assembles to bsr on architectures prior to Haswell.  Code
+   that uses lzcnt will run on older Intels and silently produce wrong
+   results.
+
+   This is a placehold. Replace with architecture-specific logic in
+   command-line options and cpuid checking. *)
+let lzcnt_support = ref false
```

Comment:
Does it really make sense to have per-instruction flags, rather than a more general architecture selection (that could impact the availability of multiple features)?

Reply:
Intel Architecture allows different combination of instruction to be supported in different CPUs and a CPU indicates which instructions it supports by setting the appropriate bit of special CPUID registers. In practice, since Haswell, existing CPUs support LZCNT. There are several target-specific  "per instruction" flags in GCC, including -mlzcnt that enables the lzcnt intrinsics to be used in the source. 
This PR suggests a different approach in which the option -flzcnt controls code generation, without the need to change the code. We can do it because the semantics of clz intrinsic proposed in PR #2179 does not have special handling for input 0, whereas GCC __builtin_clz is undefined on input 0.  
Having a target-specific per instruction flag would be useful now to enable the extra performance when it matters. Later, the flag can then be superseded by logic that checks for existing support in target CPU, either at configure time or on startup.